enclave-runner/src/usercalls/mod.rs

```diff
@@ -545,6 +557,52 @@ fn trap_attached_debugger(tcs: usize) {
     }
 }
 
+/// UsercallExtension provides a mechanism for the enclave code interface with an external service via a modified runner.
+/// If the implemenentation of the functions of UsercallExtension reutrn None, then [default functions](../../fortanix_sgx_abi/struct.Usercalls.html#networking)
+/// would be called which inteface the enclave with TCP stream.
+/// An implementation of UsercallExtension, can be registered while [building](../struct.EnclaveBuilder.html#method.usercall_ext) the enclave.
+
+pub trait UsercallExtension : 'static + Send + Sync + std::fmt::Debug {
+    /// Allows the enclave to connect to an external service specified by the addr bytes.
+    /// The Runner should determine the service that the enclave is trying to connect to by parsing the addr bytes.
+    /// If `connect_stream` returns None, the default implementation of [connect_stream](../../fortanix_sgx_abi/struct.Usercalls.html#method.connect_stream) is used.
+    /// The enclave may optionally request the local or peer addresses
+    /// be returned in `local_addr` or `peer_addr`, respectively. On success,
+    /// if `local_addr` and/or `peer_addr` is not NULL, userspace will allocate
+    /// memory for the address and populate the respective [`ByteBuffer`]
+    /// appropriately.
```

Comment:
nit: I think this is too low-level an implementation detail to put in the public API of `connect_stream`. The user of this API shouldn't have to be aware that `ByteBuffer` exists at all.

Reply:
Thanks. I guess I can drop mention about byte buffer and freeing in context of UserCallExtension, since our ABI will take care of it.